Engineering/Mobile/Flow in Kotlin.md

```diff
@@ -0,0 +1,263 @@
+
+### Context
+
+Usually, when you work with asynchronous, you must be aware of the thread you are currently running on like `Main Thread`, `Computation Thread`, or `IO Thread`. In Flow, the thread you are going to use is the thread you call the `collect` operator.
+
+**For example:**
+The snippet code below shows how can you observe a flow on the Main thread.
+```kotlin
+// MainActivity.kt
+override fun onCreate(savedInstanceState: Bundle?) {
+	// This will be executed on Dispatches.Main(Main Thread)
+	lifecycleScope.launch {
+		viewModel.userFlow.collect { user ->	
+			// ...
+		}
+	}
+}
+```
+
+But this will not be applicable if you have work that requires CPU-consuming computation or network calls. For this kind of problem, `Flow` expects you to use some operators to switch the context like:
+- `withContext` Allow you to switch the context(thread) of the flow
+    **For example:**
+    Wrapping your flow code block inside `withContext` lambda will allow it to switch from the current context to other like IO thread
+    ```kotlin
+    // In ViewModel.kt
+    val postFlow = flow {
+    	// Code block inside withContext will be executed
+    	// on Dispatchers.IO which is the IO thread
+    	withContext(Dispatchers.IO) {
+    		// ...
+    	}
+    
+    	// Outside withContext will run on the collector thread
+    }
+    
+    // Later in MainActivity.kt
+    fun fetchPosts() {
+    	lifecycleScope.launch {
+    		viewModel.postFlow.collect {
+    			// ...
+    		}
+    	}
+    }
+    ```
+    **Notes:** This approach has a limitation, that you will end up with a lot of boilerplate code.
+- `flowOn` is an operator that allows us to change the context of this flow but it's only applied to the flow before `flowOn` operator.
+    - The below example will clarify what I mean. `postFlow` itself will always run on the Background thread, but emit data to a collector on the Main thread where it being collected.
+    ```kotlin
+    val postFlow = flow {
+    	// postFlow will run on IO thread
+    }.flowOn(Dispatchers.IO)
+    
+    // Later in MainActivity.kt
+    fun fetchPosts() {
+    	// Collecting are being called on Main Thread
+    	lifecycleScope.launch {
+    		viewModel.postFlow.collect {
+    			// ...
+    		}
+    	}
+    }
+    ```
+    - You can also apply many `flowOn` operators, to perform different jobs in different contexts.
+    ```kotlin
+    val userFlow = flow {
+    	// network call on IO thread
+    }.flowOn(Dispatchers.IO)
+    .map { user ->
+    	// Execute avatar mapping on the Background thread
+    	user.avatar = "https://imageurl.com/${user.avatar}"
+    	user
+    }.flowOn(Dispatchers.Default)
+    
+    // Later in MainActivity.kt
+    fun fetchPosts() {
+    	lifecycleScope.launch {
+    		viewModel.userFlow.collect { user ->
+    			// Update UI on the Main thread
+    			updateUI(user)
+    		}
+    	}
+    }
+    ```
+    
```

Comment:
I think you have a wrong understanding of flowOn, it is just a way to switch heavy task out of main thread so it does not block the UI. Do a heavy or time-consuming task in another thread, when you have the final result to show in UI then call flowOn(Dispatchers.IO) to update UI.